awx/main/analytics/subsystem_metrics.py

```diff
@@ -0,0 +1,246 @@
+            values['counts'].append(self.buckets_to_keys[b].decode(conn))
+        values['sum'] = self.sum.decode(conn)
+        values['inf'] = self.inf.decode(conn)
+        return values
+
+    def store_value(self, conn):
+        for b in self.buckets:
+            self.buckets_to_keys[b].store_value(conn)
+        self.sum.store_value(conn)
+        self.inf.store_value(conn)
+
+    def to_prometheus(self, instance_data):
+        output_text = f"# HELP {self.field} {self.help_text}\n# TYPE {self.field} histogram\n"
+        for instance in instance_data:
+            for i, b in enumerate(self.buckets):
+                output_text += f'{self.field}_bucket{{le="{b}",node="{instance}"}} {sum(instance_data[instance][self.field]["counts"][0:i+1])}\n'
+            output_text += f'{self.field}_bucket{{le="+Inf",node="{instance}"}} {instance_data[instance][self.field]["inf"]}\n'
+            output_text += f'{self.field}_count{{node="{instance}"}} {instance_data[instance][self.field]["inf"]}\n'
+            output_text += f'{self.field}_sum{{node="{instance}"}} {instance_data[instance][self.field]["sum"]}\n'
+        return output_text
+
+
+class Metrics():
+    def __init__(self, auto_pipe_execute = True, pipe_execute_interval = 2):
+        self.pipe = redis.Redis.from_url(settings.BROKER_URL).pipeline()
+        self.conn = redis.Redis.from_url(settings.BROKER_URL)
+        self.conn.client_setname(__name__)
+        self.last_pipe_execute= time.time()
+        self.pipe_execute_interval = pipe_execute_interval
+        # auto pipe execute will commit transaction of metric data to redis
+        # at a regular interval (pipe_execute_interval). If set to False,
+        # the calling function should call .pipe_execute() explicitly
+        self.auto_pipe_execute = auto_pipe_execute
+        Instance = apps.get_model('main', 'Instance')
+        instance_name = Instance.objects.me().hostname
+        self.instance_name = instance_name
+
+        # metric name, help_text
+        METRICSLIST = [
+            SetM('callback_receiver_events_queue_size_redis',
+                 'Current number of events in redis queue'),
+            IntM('callback_receiver_events_popped_redis',
+                 'Number of events popped from redis'),
+            IntM('callback_receiver_events_in_memory',
+                 'Current number of events in memory (in transfer from redis to db)'),
+            IntM('callback_receiver_batch_events_errors',
+                 'Number of times batch insertion failed'),
+            FloatM('callback_receiver_events_insert_db_seconds',
+                   'Time spent saving events to database'),
+            IntM('callback_receiver_events_insert_db',
+                 'Number of events batch inserted into database'),
+            HistogramM('callback_receiver_batch_events_insert_db',
+                       'Number of events batch inserted into database',
+                       [10, 50, 150, 350, 650, 1000]),
```

Comment:
should this be a `setting`? i.e. batch inserts might vary depending on environment